keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1587 @@
+spec:
+  constraints:
+   rules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["deployments"]
+      scope:       "*"
+  validations:
+    # replicas is accessible because this resource matches only v1 deployments
+  - rule: "object.spec.replicas < 100"
+```
+
+Example: Typesafe access only to metadata
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "validate-xyz.example.com"
+spec:
+  constraints:
+   rules:
+    - apiGroups:   ["*"]
+      apiVersions: ["*"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["*"]
+      scope:       "*"
+  validations:
+    # minReadySeconds is not accessible because this resource matches multiple types
+  - rule: "object.spec.minReadySeconds > 60" # ERROR! Not such field "minReadySeconds".
+    # metadata is always accessible
+  - rule: "object.name.startsWith('xyz')"
+```
+
+Alternative Considered: Do not use CEL type-checking and instead dynamically evaluate
+all CEL expressions.
+
+Pros:
+
+- Can write a single CEL expression that processes multiple GVKs
+
+Cons:
+
+- All errors are reported during admission control validation. If instead
+  type-checking is used, CEL type-check errors can be surfaced much earlier in
+  the status of `ValidatingAdmissionPolicy` resources when the
+  `ValidatingAdmissionPolicy`, configuration CRDs and configuration resources
+  are written to the API Server. This provides a workflow where type-checking
+  can be used to detect errors early, preventing them from occurring during
+  admission control validation and making it safer to operate a policy using
+  `failureMode: Fail`.
+
+#### Failure Policy
+
+For in-process validation there is no remote request, so errors should be
+deterministic. We believe this significantly reduces the risk of "fail closed"
+admission control as compared to webhooks.
+
+Because failure policy is most often selected based on the need to guarantee
+enforcement, we propose defaulting failure policy to "fail" and allowing it to
+be configured on a per-rule basis:
+
+```yaml
+  validations:
+    - rule: "object.spec.xyz == configuration.x"
+      failurePolicy: Ignore # The default is "Fail"
+```
+
+TODO: Metric for ignored failures?
+
+#### Safety measures
+
+To prevent clusters from being put into a unustable state that cannot be
+recoverd from via the API, admission webhooks are not allowed to match
+`ValidatingWebhookConfiguration` and `MutatingWebhookConfiguration` kinds.
+
+We will extend this approach:
+
+- `ValidatingAdmissionPolicy` cannot match `ValidatingAdmissionPolicy`
+- `ValidatingWebhookConfiguration` cannot match `ValidatingAdmissionPolicy` or
+  `MutatingWebhookConfiguration`.
+
+Note that this does allow `ValidatingAdmissionPolicy` to match
+`ValidatingWebhookConfiguration`.
+
+We are also considering not allowing a `ValidatingAdmissionPolicy` to match the
```

Comment:
While providing tight protection, this makes hooking CRDs a pain.  I think I'd be biased towards allowing this.

Reply:
I'll drop this note since it's possible to self-rescue if a mistake is made for something like this.